File: library/src/main/java/org/incendo/interfaces/utilities/BoundInteger.java
package org.incendo.interfaces.utilities;

import org.incendo.interfaces.properties.DelegateTrigger;

public final class BoundInteger extends DelegateTrigger {

    private int value;
    private int min;
    private int max;

    public BoundInteger(final int initial, final int min, final int max) {
        this.value = initial;
        this.min = min;
        this.max = max;
    }

    public int value() {
        return this.value;
    }

    public BoundInteger value(final int newValue) {
        if (newValue >= this.min && newValue <= this.max) {
            this.value = newValue;
        } else {
            this.value = Math.min(Math.max(newValue, this.min), this.max);
        }

        this.trigger();
        return this;
    }

    public int min() {
        return this.min;
    }

    public void min(final int min) {
        this.min = min;
    }

    public int max() {
        return this.max;
    }

    public void max(final int max) {
        this.max = max;
    }

    public boolean hasSucceeding() {
        return this.value < this.max;
    }

    public boolean hasPreceding() {
        return this.value > this.min;
    }
}

File: next/src/main/java/org/incendo/interfaces/next/utilities/BoundInteger.kt
package org.incendo.interfaces.next.utilities

import org.incendo.interfaces.next.properties.DelegateTrigger
import org.incendo.interfaces.next.properties.Trigger
import kotlin.properties.ObservableProperty
import kotlin.reflect.KProperty

// todo(josh): recalculate value when max/min changed?
public class BoundInteger(
    initial: Int,
    public var min: Int,
    public var max: Int
) : ObservableProperty<Int>(initial), Trigger {

    private val delegateTrigger = DelegateTrigger()

    override fun beforeChange(property: KProperty<*>, oldValue: Int, newValue: Int): Boolean {
        val acceptableRange = min..max

        if (newValue in acceptableRange) {
            return true
        }

        val coercedValue = newValue.coerceIn(acceptableRange)
        var value by this

        value = coercedValue

        return false
    }

    override fun trigger() {
        delegateTrigger.trigger()
    }

    override fun addListener(listener: () -> Unit) {
        delegateTrigger.addListener(listener)
    }

    public fun hasSucceeding(): Boolean {
        val value by this
        return value < max
    }

    public fun hasPreceeding(): Boolean {
        val value by this
        return value > min
    }
}
